src/lib/routes/admin-api/user-admin.ts

```diff
@@ -0,0 +1,144 @@
+import Controller from '../controller';
+import { ADMIN } from '../../permissions';
+import { IUnleashConfig } from '../../types/core';
+import UserService from '../../services/user-service';
+import { AccessService, RoleName } from '../../services/access-service';
+import { Logger } from '../../logger';
+
+class UserAdminController extends Controller {
+    private userService: UserService;
+
+    private accessService: AccessService;
+
+    private logger: Logger;
+
+    constructor(config: IUnleashConfig, { userService, accessService }) {
+        super(config);
+        this.userService = userService;
+        this.accessService = accessService;
+        this.logger = config.getLogger('routes/user-controller.js');
+
+        this.get('/', this.getUsers);
+        this.get('/search', this.search);
+        this.post('/', this.createUser, ADMIN);
+        this.post('/validate-password', this.validatePassword);
+        this.put('/:id', this.updateUser, ADMIN);
+        this.post('/:id/change-password', this.changePassword, ADMIN);
+        this.delete('/:id', this.deleteUser, ADMIN);
+    }
+
+    private validateRootRole(roleName) {
+        if (!Object.values(RoleName).includes(roleName)) {
+            throw new Error(`Unknown rootRole = ${roleName}`);
+        }
+    }
+
+    async getUsers(req, res) {
+        try {
+            const users = await this.userService.getAll();
+            const rootRoles = await this.accessService.getRootRolesForAllUsers();
```

Comment:
This seems like it might have slight issues with rountrips and performance (of course it depends on how many times we're going to be making the getUsers call, but if we could write this as a single SQL query and letting our DB do the join/filtering?

Reply:
thought I was in a service or store, I see now we're in a controller, so less concerned. :)

Reply:
Your question is valid though, and I have been thinking to move this to the "user-service". The question is where should we include the root-roles (base-roles) for all users without having the frontend query each user individually. It does not belong on the user object, but it is very convenient to put it there for  the frontend / API user.